fdbserver/DeltaTree.h

```diff
@@ -1095,21 +1095,48 @@ struct DeltaTree2 {
 	// DecodedNodes are stored in a contiguous vector, which sometimes must be expanded, so care
 	// must be taken to resolve DecodedNode pointers again after the DecodeCache has new entries added.
 	struct DecodeCache : FastAllocated<DecodeCache>, ReferenceCounted<DecodeCache> {
-		DecodeCache(const T& lowerBound = T(), const T& upperBound = T())
-		  : lowerBound(arena, lowerBound), upperBound(arena, upperBound) {
+		DecodeCache(const T& lowerBound = T(), const T& upperBound = T(), size_t* pMemoryTracker = nullptr)
+		  : lowerBound(arena, lowerBound), upperBound(arena, upperBound), lastKnownUsedMemory(0),
+		    pMemoryTracker(pMemoryTracker) {
 			decodedNodes.reserve(10);
 			deltatree_printf("DecodedNode size: %d\n", sizeof(DecodedNode));
 		}
 
+		~DecodeCache() {
+			if (pMemoryTracker != nullptr) {
+				// Do not update, only subtract the last known amount which would have been
+				// published to the counter
+				*pMemoryTracker -= lastKnownUsedMemory;
+			}
+		}
+
 		Arena arena;
 		T lowerBound;
 		T upperBound;
 
+		// Track the amount of memory used by the vector and arena and publish updates to some counter.
+		// Note that no update is pushed on construction because a Cursor will surely soon follow.
+		// Updates are pushed to the counter on
+		//    DecodeCache clear
+		//    DecodeCache destruction
+		//    Cursor destruction
+		// as those are the most efficient times to publish an update.
+		size_t lastKnownUsedMemory;
+		size_t* pMemoryTracker;
+
 		// Index 0 is always the root
 		std::vector<DecodedNode> decodedNodes;
 
 		DecodedNode& get(int index) { return decodedNodes[index]; }
 
+		void updateUsedMemory() {
+			size_t usedNow = arena.getSize(true) + (decodedNodes.capacity() * sizeof(DecodedNode));
+			if (pMemoryTracker != nullptr) {
+				*pMemoryTracker += (usedNow - lastKnownUsedMemory);
+			}
+			lastKnownUsedMemory = usedNow;
```

Comment:
Given arena memory usage is an estimate and not accurate value; would that be sufficient to track 'used memory'? Or given update as you mentioned in above comment is done at selective places, it might be OK to have an expensive but more accurate measurement of total arena memory usage.

Reply:
The estimate is perfectly accurate in the DecodeCache use case.

Reply:
Which type of size calculation to do is use case dependent.  The estimate is labeled as such because of the very unusual use case described in the header and the unit test where dependencies are added to a non-root node of an arena tree.  

The boolean parameter added to `getSize()` is exactly for choosing the contract needed, but DecodeCache makes use of the knowledge that so long as `.dependsOn()` is not used the instant estimate is perfectly accurate.